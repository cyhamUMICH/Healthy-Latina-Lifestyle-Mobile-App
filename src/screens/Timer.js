import * as React from 'react';
import {Picker} from '@react-native-picker/picker'
import { View, StyleSheet, Text } from 'react-native';
import { styles } from '../styles/Styles';
import TimerView from "../components/timerview.js"
import Selectors from "../components/selectors.js"
import ActionButtons from "../components/actionbuttons.js"
//import {vibrate} from './utils'

export default class App extends React.Component {
  constructor() {
    super()
    this.state = {
      Minutes: 5,
      Seconds: 0,
      Timer: 1500,
      timer: 300,
      status: "duration",
      stopped: true,
      choosenIndex: 0,
    }
    this.beginCounter.bind(this)
  }

  beginCounter = () => {
    this.setState(prevState => ({stopped: false}))
    this.setState(prevState => ({Timer: prevState.Minutes * 60 + prevState.Seconds}))
    this.interval = setInterval(this.decrementCounter, 1000)
  }
  
  decrementCounter = () => {
    this.setState(prevState => ({timer: prevState.timer - 1}))
    if (this.state.timer === 0) {
      //vibrate()
      this.stopCounter()
    }
  }

  stopCounter = () => {
    clearInterval(this.interval)
  }

  resetCounter = () => {
    this.setState(prevState => ({timer: prevState.Timer}))
    this.setState(prevState => ({status: "duration"}))
    this.setState(prevState => ({stopped: true}))
    clearInterval(this.interval)
  }

  incrementMinutes = () => {
    if (this.state.Minutes < 59) {
      this.setState(prevState => ({Minutes: prevState.Minutes + 1}))
    }

    this.setState(prevState => ({Timer: prevState.Minutes * 60 + prevState.Seconds}))

    if (this.state.stopped) {
      this.resetCounter()
    }
  }

  incrementSeconds = () => {
    if (this.state.Seconds < 59) {
      this.setState(prevState => ({Seconds: prevState.Seconds + 1}))
    }

    this.setState(prevState => ({Timer: prevState.Minutes * 60 + prevState.Seconds}))
    
    if (this.state.stopped) {
      this.resetCounter()
    }
  }

  decrementMinutes = () => {
    if (this.state.Minutes > 0) {
      this.setState(prevState => ({Minutes: prevState.Minutes - 1}))
    }

    this.setState(prevState => ({Timer: prevState.Minutes * 60 + prevState.Seconds}))
  
    if (this.state.stopped) {
      this.resetCounter()
    }
  }

  decrementSeconds = () => {
    if (this.state.Seconds > 0) {
      this.setState(prevState => ({Seconds: prevState.Seconds - 1}))
    }

    this.setState(prevState => ({Timer: prevState.Minutes * 60 + prevState.Seconds}))
    
    if (this.state.stopped) {
      this.resetCounter()
    }
  }
  
  render() {
    return <OnlyUpdateOnEvens timer={this.state.timer} />
  }

  render() {
    
    return (
      <View style={styles.container}>
        <Picker style={styles.picker} selectedValue={this.state.language}
        onValueChange={(itemValue, itemPosition) =>  this.setState({language: itemValue, choosenIndex: itemPosition})}>
          <Picker.Item label = "Meditation" value = "meditation">
          </Picker.Item>
          <Picker.Item label = "Yoga" value = "yoga">
          </Picker.Item>
          <Picker.Item label = "Breathing" value = "breathing">
          </Picker.Item>
          <Picker.Item label = "Healing" value = "healing">
          </Picker.Item>
          <Picker.Item label = "Prayer" value = "prayer">
          </Picker.Item>
        </Picker>
        <Timer time = {this.state.timer} 
          start = {this.beginCounter} 
          reset = {this.resetCounter}
          stop = {this.stopCounter}
          minUp = {this.incrementMinutes}
          secUp = {this.incrementSeconds}
          minDw = {this.decrementMinutes}
          secDw = {this.decrementSeconds}
          minval = {this.state.Minutes}
          secval = {this.state.Seconds}
        />
      </View>
    );
  }
}

const Timer = props => {
  timeToShow = (time) => {
    let seconds = time % 60
    if (seconds < 10) {
      seconds = "0" + seconds
    }
    let minutes = (time - seconds)/60
    return minutes + ":" + seconds
  }

  return (
    <View style={styles.container}>
      <TimerView time = {this.timeToShow(props.time)} />
      <Selectors  minUp = {props.minUp}
                  minDw = {props.minDw}
                  secUp = {props.secUp}
                  secDw = {props.secDw}
                  minval = {props.minval}
                  secval = {props.secval}
                />
      <ActionButtons start = {props.start} stop = {props.stop} reset = {props.reset}/>
    </View>
  );
}