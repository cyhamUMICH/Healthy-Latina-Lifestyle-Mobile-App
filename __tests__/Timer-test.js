// Code from: https://github.com/byCedric/expo-guide-ci/blob/master/__tests__/App-test.js
//      and https://jestjs.io/docs/tutorial-react-native
import React from 'react';
import Timer from '../src/screens/Timer';
import renderer from 'react-test-renderer';

const mockedParams = {
  route: { 
    params: {
      "duration": 30,
      // In order to test the timer screen, the timer cannot be running to start.
      "isPlaying": false
    }
  }
};

// https://stackoverflow.com/questions/66563806/typeerror-cannot-read-property-params-of-undefined-jest-testing-react-nativ
test('Renders Timer With the Loading Screen', async () => {
  const tree = renderer.create(<Timer {...mockedParams} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders Timer Without the Loading Screen', async () => {
  const tree = renderer.create(<Timer {...mockedParams} skipLoadingScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});