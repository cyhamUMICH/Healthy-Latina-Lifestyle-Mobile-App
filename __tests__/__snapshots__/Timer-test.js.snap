// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Timer With the Loading Screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": "2%",
        "paddingTop": "2%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "flex": 1,
          "marginTop": "10%",
          "width": "100%",
        }
      }
    >
      <View
        accessibilityLabel="Countdown timer"
        accessible={true}
        style={
          Object {
            "height": 180,
            "position": "relative",
            "width": 180,
          }
        }
      >
        <RNSVGSvgView
          bbHeight={180}
          bbWidth={180}
          focusable={false}
          height={180}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              Object {
                "flex": 0,
                "height": 180,
                "width": 180,
              },
            ]
          }
          width={180}
        >
          <RNSVGGroup>
            <RNSVGPath
              d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
              fill={null}
              propList={
                Array [
                  "fill",
                  "stroke",
                  "strokeWidth",
                ]
              }
              stroke={4292072403}
              strokeWidth={12}
            />
            <RNSVGPath
              d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
              fill={null}
              propList={
                Array [
                  "fill",
                  "stroke",
                  "strokeWidth",
                  "strokeDasharray",
                  "strokeDashoffset",
                  "strokeLinecap",
                ]
              }
              stroke={4278208375}
              strokeDasharray={
                Array [
                  527.7875658030853,
                  527.7875658030853,
                ]
              }
              strokeDashoffset={null}
              strokeLinecap={1}
              strokeWidth={12}
            />
          </RNSVGGroup>
        </RNSVGSvgView>
        <View
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "height": "100%",
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "rgba(0, 71, 119, 1)",
                "fontSize": 30,
              }
            }
          >
            00:00:30
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "flexDirection": "row",
            "marginTop": "10%",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "aspectRatio": 1,
              "backgroundColor": "#00A89D",
              "borderRadius": 667,
              "height": null,
              "marginHorizontal": "5%",
              "opacity": 1,
              "width": "15%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "aspectRatio": 1,
              "backgroundColor": "#00A89D",
              "borderRadius": 667,
              "height": null,
              "marginHorizontal": "5%",
              "opacity": 1,
              "width": "15%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "width": "100%",
            },
            Object {
              "zIndex": 5000,
            },
            Object {
              "marginTop": "5%",
            },
            false,
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          forwardedRef={[Function]}
          onClick={[Function]}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderColor": "#000",
              "borderRadius": 8,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingHorizontal": 10,
              "width": "100%",
              "zIndex": 5000,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "flex": 1,
                },
                Object {
                  "fontSize": 18,
                  "textAlign": "left",
                },
                false,
                Object {},
              ]
            }
          >
            Select an intention...
          </Text>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 10,
                },
                Object {},
              ]
            }
          >
            <Image
              source={1}
              style={
                Array [
                  Object {
                    "height": 20,
                    "width": 20,
                  },
                  Object {},
                ]
              }
            />
          </View>
        </View>
      </View>
      <RCTScrollView
        style={
          Object {
            "alignSelf": "flex-start",
            "marginVertical": "2%",
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "fontSize": 18,
                "textAlign": "left",
              }
            }
          />
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;

exports[`Renders Timer Without the Loading Screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": "2%",
        "paddingTop": "2%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "flex": 1,
          "marginTop": "10%",
          "width": "100%",
        }
      }
    >
      <View
        accessibilityLabel="Countdown timer"
        accessible={true}
        style={
          Object {
            "height": 180,
            "position": "relative",
            "width": 180,
          }
        }
      >
        <RNSVGSvgView
          bbHeight={180}
          bbWidth={180}
          focusable={false}
          height={180}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              Object {
                "flex": 0,
                "height": 180,
                "width": 180,
              },
            ]
          }
          width={180}
        >
          <RNSVGGroup>
            <RNSVGPath
              d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
              fill={null}
              propList={
                Array [
                  "fill",
                  "stroke",
                  "strokeWidth",
                ]
              }
              stroke={4292072403}
              strokeWidth={12}
            />
            <RNSVGPath
              d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
              fill={null}
              propList={
                Array [
                  "fill",
                  "stroke",
                  "strokeWidth",
                  "strokeDasharray",
                  "strokeDashoffset",
                  "strokeLinecap",
                ]
              }
              stroke={4278208375}
              strokeDasharray={
                Array [
                  527.7875658030853,
                  527.7875658030853,
                ]
              }
              strokeDashoffset={null}
              strokeLinecap={1}
              strokeWidth={12}
            />
          </RNSVGGroup>
        </RNSVGSvgView>
        <View
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "height": "100%",
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "rgba(0, 71, 119, 1)",
                "fontSize": 30,
              }
            }
          >
            00:00:30
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "flexDirection": "row",
            "marginTop": "10%",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "aspectRatio": 1,
              "backgroundColor": "#00A89D",
              "borderRadius": 667,
              "height": null,
              "marginHorizontal": "5%",
              "opacity": 1,
              "width": "15%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "aspectRatio": 1,
              "backgroundColor": "#00A89D",
              "borderRadius": 667,
              "height": null,
              "marginHorizontal": "5%",
              "opacity": 1,
              "width": "15%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "width": "100%",
            },
            Object {
              "zIndex": 5000,
            },
            Object {
              "marginTop": "5%",
            },
            false,
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          forwardedRef={[Function]}
          onClick={[Function]}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderColor": "#000",
              "borderRadius": 8,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingHorizontal": 10,
              "width": "100%",
              "zIndex": 5000,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "flex": 1,
                },
                Object {
                  "fontSize": 18,
                  "textAlign": "left",
                },
                false,
                Object {},
              ]
            }
          >
            Select an intention...
          </Text>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 10,
                },
                Object {},
              ]
            }
          >
            <Image
              source={1}
              style={
                Array [
                  Object {
                    "height": 20,
                    "width": 20,
                  },
                  Object {},
                ]
              }
            />
          </View>
        </View>
      </View>
      <RCTScrollView
        style={
          Object {
            "alignSelf": "flex-start",
            "marginVertical": "2%",
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "fontSize": 18,
                "textAlign": "left",
              }
            }
          />
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
